File: immersion/src/contextmenu.rs
//! The context menu: right-click an area for its operations.
//!
//! The menu itself lives in the DOM — each right-clickable element carries its
//! items as a `data-im-menu` attribute — and the shim ([`contextmenu.js`]) draws
//! and drives it entirely client-side. This component only opens the eval
//! channel the shim commits a pick over, and hands each pick to the host as
//! `(action, params)` — the same shape a button or a chord produces, so a
//! right-click Split and a header-button Split are one operation on one bus.
//!
//! Renders nothing; place it once, like [`crate::Keymap`].

use dioxus::prelude::*;
use serde_json::Value;

const CONTEXTMENU_JS: &str = include_str!("contextmenu.js");

/// One row of a menu. Menus were assembled as JSON text, which nothing on this
/// side checked and which broke the moment a label carried a quote — a run id
/// or a workspace name would have been enough. They are values now, serialised
/// by serde, and the shim's `MenuItem` is generated from this rather than
/// written to match it.
#[derive(Debug, Clone, Default, serde::Serialize, ts_rs::TS)]
#[ts(export, export_to = "../ts/generated/")]
pub struct MenuItem {
    // `optional` so the generated type says `label?: string` — matching what
    // serde actually emits, since these are skipped when absent. Without it the
    // TypeScript would promise a key that is not there.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub label: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub action: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional, type = "unknown")]
    pub params: Option<Value>,
    /// The chord to show right-aligned, already written the platform's way.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub chord: Option<String>,
    /// Inline SVG shown before the label. The shim inserts it as markup, so
    /// it is the library's own sprite output and never user text.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[ts(optional)]
    pub icon: Option<String>,
    /// A divider rather than a row. Always emitted, so the type can promise it.
    pub sep: bool,
    /// Shown, but not runnable — Blender greys an operator whose `poll` fails
    /// rather than hiding it, because a row that disappears teaches nothing
    /// and a row that is there but dim says "this exists, not now".
    pub disabled: bool,
}

impl MenuItem {
    /// A row that runs `action` with `params`.
    pub fn new(label: &str, action: &str, params: Value) -> Self {
        MenuItem {
            label: Some(label.to_string()),
            action: Some(action.to_string()),
            params: Some(params),
            ..Default::default()
        }
    }

    /// Show a chord beside the label.
    /// Show an icon before the label. Takes a sprite name, not markup, so a
    /// caller cannot put arbitrary HTML into a menu.
    pub fn with_icon(mut self, name: &str) -> Self {
        let svg = crate::icons::icon(name);
        self.icon = (!svg.is_empty()).then_some(svg);
        self
    }

    pub fn with_chord(mut self, chord: &str) -> Self {
        self.chord = Some(chord.to_string());
        self
    }

    /// Grey this row out unless `available`. Takes the answer rather than the
    /// question, because only the host holds the registry the answer comes
    /// from — `Commands::can`.
    pub fn when(mut self, available: bool) -> Self {
        self.disabled = !available;
        self
    }

    /// A divider.
    pub fn sep() -> Self {
        MenuItem {
            sep: true,
            ..Default::default()
        }
    }
}

/// Serialise a menu for a `data-im-menu` attribute. Escaping is serde's problem
/// now, not the caller's.
pub fn menu_json(items: &[MenuItem]) -> String {
    serde_json::to_string(items).unwrap_or_else(|_| "[]".into())
}

#[derive(serde::Deserialize, ts_rs::TS)]
#[ts(export, export_to = "../ts/generated/", rename = "MenuPick")]
pub struct Pick {
    action: String,
    /// Whatever the menu item carried. `unknown` on the TypeScript side, which
    /// is honest: a pick's params are the command's business, not the shim's.
    #[ts(type = "unknown")]
    params: Value,
}

#[derive(Props, Clone)]
pub struct ContextMenuProps {
    /// `(action, params)` for a picked item — routed like any command.
    pub on_command: Callback<(String, Value)>,
}

impl PartialEq for ContextMenuProps {
    fn eq(&self, _: &Self) -> bool {
        true
    }
}

/// Installs the context-menu shim and routes picks to the host.
#[component]
pub fn ContextMenu(props: ContextMenuProps) -> Element {
    let on_command = props.on_command;
    use_future(move || async move {
        let mut channel = dioxus::document::eval(CONTEXTMENU_JS);
        loop {
            let Ok(raw) = channel.recv::<String>().await else {
                return; // channel closed; a reload re-installs
            };
            if let Ok(pick) = serde_json::from_str::<Pick>(&raw) {
                on_command.call((pick.action, pick.params));
            }
        }
    });
    rsx! {}
}

/// The editor switcher's menu — every registered editor kind, as a click-open
/// list. Blender's leftmost header button is a menu of editor types, not a
/// dropdown control, and a menu can show labels the way the registry names them.
pub fn editor_menu_json(id: crate::AreaId, kinds: &[crate::EditorKind], current: &str) -> String {
    let items: Vec<MenuItem> = kinds
        .iter()
        .map(|k| {
            let mark = if k.id == current { "• " } else { "" };
            MenuItem::new(
                &format!("{mark}{}", k.label),
                "set_editor",
                serde_json::json!({ "id": id, "editor": k.id }),
            )
            .with_icon(k.icon)
        })
        .collect();
    menu_json(&items)
}

/// The area header's View dropdown — region toggles and the area operations,
/// Blender's per-editor View menu. Click-opened (data-im-menu-click), so it
/// reads as a menu button rather than a right-click surprise.
pub fn view_menu_json(id: crate::AreaId, toolbar: bool, sidebar: bool, regions: bool) -> String {
    let mut items = Vec::new();
    if regions {
        let word = |on: bool| if on { "Hide" } else { "Show" };
        items.push(
            MenuItem::new(
                &format!("{} Toolbar", word(toolbar)),
                "toggle_region",
                serde_json::json!({ "id": id, "region": crate::area::Region::Toolbar.as_str() }),
            )
            .with_chord("T"),
        );
        items.push(
            MenuItem::new(
                &format!("{} Sidebar", word(sidebar)),
                "toggle_region",
                serde_json::json!({ "id": id, "region": crate::area::Region::Sidebar.as_str() }),
            )
            .with_chord("N"),
        );
        items.push(MenuItem::sep());
    }
    items.extend([
        MenuItem::new(
            "Split horizontal",
            "split",
            serde_json::json!({ "id": id, "dir": "row" }),
        ),
        MenuItem::new(
            "Split vertical",
            "split",
            serde_json::json!({ "id": id, "dir": "col" }),
        ),
        MenuItem::new(
            "Duplicate",
            "duplicate_area",
            serde_json::json!({ "id": id }),
        ),
        MenuItem::sep(),
        MenuItem::new(
            "Hide header",
            "toggle_region",
            serde_json::json!({ "id": id, "region": crate::area::Region::Header.as_str() }),
        ),
        MenuItem::new(
            "Flip header",
            "toggle_region",
            serde_json::json!({ "id": id, "region": crate::area::Region::HeaderFlip.as_str() }),
        ),
        MenuItem::sep(),
        MenuItem::new("Close area", "join", serde_json::json!({ "id": id })),
    ]);
    menu_json(&items)
}

/// The `data-im-menu` JSON for an area leaf — split either way, then close.
/// Kept here so the area view and the menu never drift: both come from the id.
pub fn area_menu_json(id: crate::AreaId, can: impl Fn(&str) -> bool) -> String {
    menu_json(&[
        MenuItem::new(
            "Split horizontal",
            "split",
            serde_json::json!({ "id": id, "dir": "row" }),
        )
        .when(can("split")),
        MenuItem::new(
            "Split vertical",
            "split",
            serde_json::json!({ "id": id, "dir": "col" }),
        )
        .when(can("split")),
        MenuItem::new(
            "Duplicate",
            "duplicate_area",
            serde_json::json!({ "id": id }),
        )
        .when(can("duplicate_area")),
        MenuItem::sep(),
        // The one everybody meets: the last area has nothing to join into, so
        // this has always been a row that looked live and did nothing.
        MenuItem::new("Close area", "join", serde_json::json!({ "id": id })).when(can("join")),
    ])
}

File: immersion/ts/contextmenu.ts
// The menu shim: Blender's right-click menus, its click-open dropdowns, and
// its pie menus, client-side.
//
// An element carries its menu as a JSON array of MenuItem in one of two
// attributes: data-im-menu opens on right-click at the cursor, and
// data-im-menu-click opens on left-click anchored under the element. A pie is
// raised by the host through __imOpenPie. Either way the shim draws and drives
// the menu, and commits ONE message — the picked {action, params}.

import { once, send, type MenuItem, type MenuPick } from "./types";

if (once("__imCtxMenu")) {

  let menu: HTMLElement | null = null;
  let onKey: ((e: KeyboardEvent) => void) | null = null;
  // Where the pointer last was — Blender's Q opens the favourites at the mouse.
  let lastX = 0;
  let lastY = 0;
  document.addEventListener("pointermove", (e) => { lastX = e.clientX; lastY = e.clientY; }, true);
  let openFor: HTMLElement | null = null; // the element whose dropdown is open, for toggle-off

  const close = () => {
    if (menu) {
      menu.remove();
      menu = null;
    }
    if (onKey) {
      document.removeEventListener("keydown", onKey, true);
      onKey = null;
    }
    openFor = null;
  };

  const commit = (action: string, params: unknown): void =>
    send({ action, params: params ?? null } satisfies MenuPick);

  // Copy / Paste value are clipboard operations — handled here, not sent to the
  // server (which has no clipboard). dioxus.send only works from a synchronous
  // handler, not a detached promise, so paste cannot read-then-send inside the
  // click; instead the clipboard is prefetched when the menu opens and the pick
  // sends the stashed value synchronously.
  let clipStash: string | null = null;
  const pick = (action: string, paramsStr: string | undefined): void => {
    const params = JSON.parse(paramsStr || "null") as Record<string, unknown> | null;
    if (action === "copy_value") {
      const v = params?.["value"];
      navigator.clipboard
        .writeText(typeof v === "string" ? v : JSON.stringify(v))
        .catch(() => {});
      return;
    }
    if (action === "paste_value") {
      if (clipStash == null) return;
      let v;
      try {
        v = JSON.parse(clipStash);
      } catch (_) {
        v = clipStash;
      }
      commit("set_setting", { pointer: params?.["pointer"], value: v });
      return;
    }
    commit(action, params);
  };

  // Build and place the menu. `at` is either {x, y} (cursor) or {rect} (anchor
  // the menu under an element's box, the dropdown case).
  const openMenu = (items: MenuItem[], at: { x?: number; y?: number; rect?: DOMRect }): void => {
    close();
    if (items.some((it) => it.action === "paste_value")) {
      navigator.clipboard.readText().then((t) => { clipStash = t; }).catch(() => {});
    }

    menu = document.createElement("div");
    menu.className = "im-ctxmenu";
    for (const it of items) {
      if (it.sep) {
        const s = document.createElement("div");
        s.className = "im-ctx-sep";
        menu.appendChild(s);
        continue;
      }
      const row = document.createElement("div");
      // A row whose command cannot run right now is dim and inert, not
      // missing: a row that disappears teaches nothing, and one that is there
      // but grey says "this exists, not now".
      row.className = it.disabled ? "im-ctx-item is-disabled" : "im-ctx-item";
      if (it.icon) {
        // The icon is the library's own sprite output — markup by
        // construction, never user text — so it is inserted as markup while
        // the label stays textContent.
        const glyph = document.createElement("span");
        glyph.className = "im-ctx-icon";
        glyph.innerHTML = it.icon;
        row.appendChild(glyph);
      }
      const label = document.createElement("span");
      label.textContent = it.label ?? "";
      row.appendChild(label);
      if (it.chord) {
        const c = document.createElement("span");
        c.className = "im-ctx-chord";
        c.textContent = it.chord;
        row.appendChild(c);
      }
      row.dataset.action = it.action ?? "";
      row.dataset.params = JSON.stringify(it.params ?? null);
      row.addEventListener("click", () => {
        if (it.disabled) return;
        pick(row.dataset.action ?? "", row.dataset.params);
        close();
      });
      // Blender's "Add to Quick Favourites": right-click a menu row.
      row.addEventListener("contextmenu", (ev) => {
        ev.preventDefault();
        ev.stopPropagation();
        commit("favorite_add", {
          label: it.label,
          action: it.action,
          params: it.params ?? null,
        });
        close();
      });
      menu.appendChild(row);
    }
    document.body.appendChild(menu);

    const w = menu.offsetWidth;
    const h = menu.offsetHeight;
    let x: number;
    let y: number;
    if (at.rect) {
      x = at.rect.left;
      y = at.rect.bottom + 2; // hang under the button
      if (y + h > window.innerHeight - 4) y = at.rect.top - h - 2; // flip above
    } else {
      x = at.x ?? 0;
      y = at.y ?? 0;
      if (y + h > window.innerHeight - 4) y = window.innerHeight - h - 4;
    }
    if (x + w > window.innerWidth - 4) x = window.innerWidth - w - 4;
    menu.style.left = Math.max(4, x) + "px";
    menu.style.top = Math.max(4, y) + "px";

    const el = menu;
    // Disabled rows are excluded here rather than only from the click
    // handler: this list is what the arrows walk, what Enter fires and what
    // the letter accelerators match, so leaving them in would let the
    // keyboard run a command the pointer refuses.
    const rows = (): HTMLElement[] =>
      Array.from(el.querySelectorAll<HTMLElement>(".im-ctx-item:not(.is-disabled)"));
    let sel = 0;
    const paint = () => rows().forEach((r, i) => r.classList.toggle("is-sel", i === sel));
    paint();

    onKey = (ev) => {
      const rs = rows();
      if (ev.key === "ArrowDown") {
        ev.preventDefault();
        sel = (sel + 1) % rs.length;
        paint();
      } else if (ev.key === "ArrowUp") {
        ev.preventDefault();
        sel = (sel - 1 + rs.length) % rs.length;
        paint();
      } else if (ev.key === "Enter") {
        ev.preventDefault();
        const r = rs[sel];
        if (r) pick(r.dataset.action ?? "", r.dataset.params);
        close();
      } else if (ev.key === "Escape") {
        ev.preventDefault();
        close();
      } else if (ev.key.length === 1 && /[a-z0-9]/i.test(ev.key)) {
        // Accelerators: a letter jumps to the first item starting with it, and
        // picks straight away when only one matches — the way a desktop menu
        // has always worked. Repeating the letter cycles the matches.
        const want = ev.key.toLowerCase();
        const hits = rs.filter((r) => (r.textContent ?? "").trim().toLowerCase().startsWith(want));
        if (hits.length === 0) return;
        ev.preventDefault();
        if (hits.length === 1) {
          const only = hits[0];
          if (only) pick(only.dataset.action ?? "", only.dataset.params);
          close();
          return;
        }
        const cur = rs[sel];
        const from = cur && hits.includes(cur) ? hits.indexOf(cur) + 1 : 0;
        const next = hits[from % hits.length];
        if (next) sel = rs.indexOf(next);
        paint();
      }
    };
    document.addEventListener("keydown", onKey, true);
  };

  const parse = (raw: string | undefined): MenuItem[] | null => {
    try {
      const items = JSON.parse(raw ?? "") as MenuItem[];
      return Array.isArray(items) && items.length > 0 ? items : null;
    } catch (_) {
      return null;
    }
  };

  // Right-click: a context menu at the cursor.
  document.addEventListener("contextmenu", (e) => {
    const host = (e.target as Element | null)?.closest?.<HTMLElement>("[data-im-menu]");
    if (!host) return;
    const items = parse(host.dataset.imMenu);
    if (!items) return;
    e.preventDefault();
    openMenu(items, { x: e.clientX, y: e.clientY });
  });

  // Left-click: a dropdown anchored under the button. Clicking the open
  // button again closes it, the way a menu bar behaves.
  document.addEventListener("click", (e) => {
    const host = (e.target as Element | null)?.closest?.<HTMLElement>("[data-im-menu-click]");
    if (!host) return;
    e.preventDefault();
    e.stopPropagation();
    if (openFor === host) {
      close();
      return;
    }
    const items = parse(host.dataset.imMenuClick);
    if (!items) return;
    openMenu(items, { rect: host.getBoundingClientRect() });
    openFor = host;
  });

  // A menu the host raises from a keypress rather than a click — Quick
  // Favourites (Q). The shim's own channel carries the pick, so the host only
  // has to hand over the items.
  window.__imOpenMenu = (itemsJson: string): void => {
    const items = parse(itemsJson);
    if (!items) return;
    openMenu(items, { x: lastX || 40, y: lastY || 40 });
  };


  // --- pie menu -------------------------------------------------------------
  // Blender's radial menu: slices arranged around the pointer, the one nearest
  // the cursor highlighted, picked by click (or by releasing a drag out onto
  // it). Up to eight slices, ordered the way Blender lays them out — west,
  // east, south, north, then the diagonals — so muscle memory has somewhere to
  // form. Params carrying the string "@area" are resolved to the id of the area
  // under the pointer, so one pie definition works for whichever area you are
  // over.
  const PIE_ANGLES = [180, 0, 90, 270, 135, 45, 225, 315];
  const PIE_R = 92;
  let pie: { el: HTMLElement; slices: { el: HTMLElement; x: number; y: number }[]; cx: number; cy: number } | null = null;

  const resolveArea = (params: unknown, areaId: number | null): unknown => {
    if (params == null || typeof params !== "object") return params;
    const out: Record<string, unknown> = Array.isArray(params) ? ([] as never) : {};
    for (const k of Object.keys(params as Record<string, unknown>)) {
      const v = (params as Record<string, unknown>)[k];
      out[k] = v === "@area" ? areaId : (typeof v === "object" ? resolveArea(v, areaId) : v);
    }
    return out;
  };

  const closePie = () => {
    if (pie) {
      pie.el.remove();
      pie = null;
    }
    if (onKey) {
      document.removeEventListener("keydown", onKey, true);
      onKey = null;
    }
  };

  window.__imOpenPie = (itemsJson: string): void => {
    const items = parse(itemsJson);
    if (!items) return;
    close();
    closePie();
    const cx = lastX || window.innerWidth / 2;
    const cy = lastY || window.innerHeight / 2;
    const under = document.elementFromPoint(cx, cy);
    const areaEl = under?.closest ? under.closest<HTMLElement>(".im-area") : null;
    const areaId = areaEl ? Number((areaEl as HTMLElement).dataset.imArea) : null;

    const el = document.createElement("div");
    el.className = "im-pie";
    const slices: { el: HTMLElement; x: number; y: number }[] = [];
    items.slice(0, 8).forEach((it, i) => {
      const ang = ((PIE_ANGLES[i] ?? 0) * Math.PI) / 180;
      const x = cx + Math.cos(ang) * PIE_R;
      const y = cy + Math.sin(ang) * PIE_R;
      const s = document.createElement("div");
      s.className = "im-pie-slice";
      s.textContent = it.label ?? "";
      s.style.left = x + "px";
      s.style.top = y + "px";
      s.addEventListener("click", () => {
        pick(it.action ?? "", JSON.stringify(resolveArea(it.params ?? null, areaId)));
        closePie();
      });
      el.appendChild(s);
      slices.push({ el: s, x, y });
    });
    const dot = document.createElement("div");
    dot.className = "im-pie-center";
    dot.style.left = cx + "px";
    dot.style.top = cy + "px";
    el.appendChild(dot);
    document.body.appendChild(el);
    pie = { el, slices, cx, cy };

    // Highlight whichever slice the pointer is nearest.
    const track = (e: PointerEvent): void => {
      if (!pie) return;
      let best: { el: HTMLElement; x: number; y: number } | null = null;
      let bestD = Infinity;
      for (const s of pie.slices) {
        const d = Math.hypot(e.clientX - s.x, e.clientY - s.y);
        if (d < bestD) {
          bestD = d;
          best = s;
        }
      }
      for (const s of pie.slices) s.el.classList.toggle("is-sel", s === best && bestD < 140);
    };
    el.addEventListener("pointermove", track);
    document.addEventListener("pointermove", track, true);

    onKey = (ev) => {
      if (ev.key === "Escape") {
        ev.preventDefault();
        closePie();
      }
    };
    document.addEventListener("keydown", onKey, true);
    el.addEventListener("pointerdown", (e) => {
      // A press on the backdrop (not a slice) dismisses.
      if (e.target === el) closePie();
    });
  };

  // Dismiss on any outside press.
  document.addEventListener(
    "pointerdown",
    (e) => {
      const t = e.target as Element | null;
      if (menu && t && !menu.contains(t) && !t.closest?.("[data-im-menu-click]")) {
        close();
      }
    },
    true,
  );
}
